pkg/util/watch.go

```diff
@@ -511,3 +549,127 @@ func (w *WatchStore[T]) Items() []*T {
 
 	return items
 }
+
+// NewIndexedWatchStore creates a new IndexedWatchStore from the WatchStore and the index to use.
+//
+// Note: the index type is assumed to have reference semantics; i.e. any shallow copy of the value
+// will affect any other shallow copy.
+//
+// For more information, refer to IndexedWatchStore.
+func NewIndexedWatchStore[T any, I WatchIndex[T]](store *WatchStore[T], index I) IndexedWatchStore[T, I] {
+	store.mutex.Lock()
+	defer store.mutex.Unlock()
+
+	for _, obj := range store.objects {
+		index.Add(obj)
+	}
+
+	id := store.nextIndexID
+	store.nextIndexID += 1
+
+	collector := &struct{}{}
+	// when this IndexedWatchStore is GC'd, remove its index from the WatchStore. This should
+	// provide a reliable way of making sure that indexes always get cleaned up.
+	stdruntime.SetFinalizer(collector, func(_ any) {
+		// note: finalizers always run in a separate goroutine, so it's ok to lock here.
+		store.mutex.Lock()
+		defer store.mutex.Unlock()
+		delete(store.indexes, id)
+	})
+
+	return IndexedWatchStore[T, I]{store, index, id, collector}
+}
+
+// IndexedWatchStore represents a WatchStore, wrapped with a privileged WatchIndex that can be used
+// to efficiently answer queries.
+type IndexedWatchStore[T any, I WatchIndex[T]] struct {
+	*WatchStore[T]
+
+	index I
+
+	// id stores the id of this index in the WatchStore
+	id uint64
+	// collector allows us to
+	collector *struct{}
+}
+
+// WithIndex calls a function with the current state of the index, locking the WatchStore around it.
+//
+// It is almost guaranteed to be an error to indirectly return the index with this function.
+func (w IndexedWatchStore[T, I]) WithIndex(f func(I)) {
+	w.WatchStore.mutex.Lock()
+	defer w.WatchStore.mutex.Unlock()
+
+	f(w.index)
+}
+
+func (w IndexedWatchStore[T, I]) GetIndexed(f func(I) (*T, bool)) (obj *T, ok bool) {
+	w.WithIndex(func(i I) {
+		obj, ok = f(i)
+	})
+	return
+}
+
+func (w IndexedWatchStore[T, I]) ListIndexed(f func(I) []*T) (list []*T) {
+	w.WithIndex(func(i I) {
+		list = f(i)
+	})
+	return
+}
+
+func NewNameIndex[T any]() *NameIndex[T] {
+	// check that *T implements metav1.ObjectMetaAccessor
+	var zero T
+	ptrToZero := any(&zero)
+	if _, ok := ptrToZero.(metav1.ObjectMetaAccessor); !ok {
+		panic("type *T must implement metav1.ObjectMetaAccessor")
+	}
+
+	// FIXME: does this need to be a pointer?
+	return &NameIndex[T]{
+		namespacedNames: make(map[struct {
+			namespace string
+			name      string
+		}]*T),
+	}
+}
+
+// NameIndex is a WatchIndex that provides efficient lookup for a value with a particular name
+type NameIndex[T any] struct {
+	namespacedNames map[struct {
+		namespace string
+		name      string
+	}]*T
+}
+
+// note: requires that *T implements metav1.ObjectMetaAccessor
+func keyForObj[T any](obj *T) struct {
+	namespace string
+	name      string
+} {
+	meta := any(obj).(metav1.ObjectMetaAccessor).GetObjectMeta()
```

Comment:
I really want there to be a way to specialize/type using interface types rather than generics, given that we have to do this runtime cast anyway, but I suspect you've had the same feeling and have figured out that there isn't

Reply:
Yeah, basically what's going on here is that the `WatchStore` is parameterized by the type `T` rather than `*T`, so its indexes must be as well. The issue is that `metav1.ObjectMetaAccessor` is generally implemented for `*T`, not `T`, so we can't directly require that for this index's type without adding a second generic parameter (which would be too much of a hassle to deal with).